### PR TITLE
Migrate tests to cucumber

### LINKED ISSUE #65: test: migrate integration test to cucumber
Our integration tests are using bespoke lifecycle management which isn't great for maintainability. [cucuber-rs](https://github.com/cucumber-rs/cucumber) provides lifecycle hooks and a behavioural driven approach which works well for integration tests.

We should explore replacing the bespoke lifecycle-managed tests with cucumber.

### PR DESCRIPTION
**Pull request by ZoopOTheGoop**
As the title says, this will migrate the integration tests (or some subset) to Cucumber, currently under investigation.

Closes #65

### CODE REVIEW

**Comment by ZoopOTheGoop** [comment]
This is more or less done. A few things should probably be renamed for consistency.

Unlike the old integration tests, the new cucumber tests really only test externally observable behavior (which I think is a strength). The *one* exception is double authentication. It's very hard to meaningfully tell the difference between double auths without just asking the DB if someone got added twice somehow, so that's the only bit of test SQL.

One side thing is because of the high volume of votes for testing categories, i ended up having to parallelize the vote generator. This is a little ugly, but I think it's for the best.

**Review by tim-hm — CHANGES_REQUESTED** [review]
This is good work @ZoopOTheGoop. Just a few tweaks related to more how this would be used in practice please:

- In your givens, you could do a quick check to see if infra is running correctly and bail with a message suggesting how the tests should be run (eg `cargo make full-test` or something else)
- If you want to stick with `cargo-make`, do you think its worth adding as a build dep?
- Test infra should ideally be different from the infra used for local development (ie uses a different sql db, runs on different ports, etc). The goal is someone should be able to run the server locally & be able to run tests simultaneously without them tainting each other. Can you explore spinning up 'test' infra distinct from the infra used to iterate locally? Happy for that work to be in another PR.
- We should be able to run the entire test suite with a single command (ie at the moment its: `cargo make db-up && cargo make full-test`. 

Finally, when I run the test suite locally, using `cargo make db-up && cargo make full-test` the suite gets stuck on "List of top 20 staps'. Does it succeed on your end? This is the top of the output I get with clean infra which I have to ctrl+c to stop:

```
Feature: List of top 20 snaps
  Scenario: Tails opens the store homepage, seeing the top snaps
   ✔> Given a snap with id "3Iwi803Tk3KQwyD6jFiAJdlq8MLgBIoD" gets 100 votes where 75 are upvotes
    > Given 25 test snaps gets between 150 and 200 votes, where 125 to 175 are upvotes
2024-02-21T22:05:16.153301Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
2024-02-21T22:05:16.156976Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
2024-02-21T22:05:16.207071Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
2024-02-21T22:05:16.207803Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
2024-02-21T22:05:16.229790Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
2024-02-21T22:05:16.238972Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
```

**Comment by ZoopOTheGoop** [comment]
> This is good work @ZoopOTheGoop. Just a few tweaks related to more how this would be used in practice please:
> 
>     * In your givens, you could do a quick check to see if infra is running correctly and bail with a message suggesting how the tests should be run (eg `cargo make full-test` or something else)
> 
>     * If you want to stick with `cargo-make`, do you think its worth adding as a build dep?
> 
>     * Test infra should ideally be different from the infra used for local development (ie uses a different sql db, runs on different ports, etc). The goal is someone should be able to run the server locally & be able to run tests simultaneously without them tainting each other. Can you explore spinning up 'test' infra distinct from the infra used to iterate locally? Happy for that work to be in another PR.
> 
>     * We should be able to run the entire test suite with a single command (ie at the moment its: `cargo make db-up && cargo make full-test`.
> 
> 
> Finally, when I run the test suite locally, using `cargo make db-up && cargo make full-test` the suite gets stuck on "List of top 20 staps'. Does it succeed on your end? This is the top of the output I get with clean infra which I have to ctrl+c to stop:
> 
> ```
> Feature: List of top 20 snaps
>   Scenario: Tails opens the store homepage, seeing the top snaps
>    ✔> Given a snap with id "3Iwi803Tk3KQwyD6jFiAJdlq8MLgBIoD" gets 100 votes where 75 are upvotes
>     > Given 25 test snaps gets between 150 and 200 votes, where 125 to 175 are upvotes
> 2024-02-21T22:05:16.153301Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
> 2024-02-21T22:05:16.156976Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
> 2024-02-21T22:05:16.207071Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
> 2024-02-21T22:05:16.207803Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
> 2024-02-21T22:05:16.229790Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
> 2024-02-21T22:05:16.238972Z  WARN ratings::features::user::use_cases: an error occurred when calling snapd: an error happened when attempting to retrieve info via /assertions/snap-declaration didn't find a snap with the given id
> ```

You have to be *very* patient with that test. If you want to see progress change the logging level to `Debug`. It takes a hot minute to run because of all the requests.

The warnings are going to happen no matter what, since we're randomly generating Snap IDs, any request to `snapd` to look it up is going to fail, because they're not real IDs. The actual program is set up to handle this (hence why it's only at `Warn` level, it's a problem in production but expected in tests). I'd recommend changing the log level to `Error` for just the server or else running the actual server in another tab. This happened on the old integration tests as well after introducing the category stuff in #70.

File: src/features/common/entities.rs [hunk 1]
```diff
@@ -45,10 +47,38 @@ impl RatingsBand {
...
 
+impl PartialOrd for RatingsBand {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        if matches!(self, RatingsBand::InsufficientVotes)
+            || matches!(other, RatingsBand::InsufficientVotes)
+        {
+            None
+        } else {
+            // The ratings bad values are actually backwards from an ordering perspective,
+            // so flip them
```
[the remark this answers is not in the document]
Comment (ZoopOTheGoop) [reply src/features/common/entities.rs:58]: Yeah, I went to change it but wanted to see how/if it impacted the snap store first. I'll open an issue.

Will change the comment though.

**Comment by ZoopOTheGoop** [comment]
Re: infra, I don't think it's good as a `given`, but could just be put in `main` before the tests as a helper function.

Re: `cargo-make` I don't actually use it. I've been running with three tabs: `dotenv -f test.env podman-compose up`; `dotenv -f test.env cargo run --release`; and `cargo test --release`.

File: src/features/common/entities.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tim-hm) [reply src/features/common/entities.rs:58]: This is another instance where the ACL approach is helpful ... ie maintain internal consistency by adjusting structures/types at our boundaries

**Comment by tim-hm** [comment]
Thanks for the responses. 

Agreed main sounds like a better place to checking if the infra is ready.

For cargo-make my focus there is for other people who are new to the project and aren't as familiar with how to get everything up and running. The idea being we can say here's a single command to run the test suite or here's a single command to iterate locally. We've provided the default happy path and then individuals can choose to go off piste.

**Review by tim-hm — APPROVED** [review]
Feel free to merge when you're ready

**Pull request merged**